src/validation.cpp

```diff
@@ -4592,16 +4592,20 @@ bool LoadGenesisBlock(const CChainParams& chainparams)
     return ::ChainstateActive().LoadGenesisBlock(chainparams);
 }
 
-bool LoadExternalBlockFile(const CChainParams& chainparams, FILE* fileIn, FlatFilePos *dbp)
+bool LoadExternalBlockFile(const CChainParams& chainparams, CBufferedFile& blkdat, FlatFilePos *dbp)
 {
     // Map of disk positions for blocks with unknown parent (only used for reindex)
-    static std::multimap<uint256, FlatFilePos> mapBlocksUnknownParent;
+    struct unknownParentPos {
+        FlatFilePos filePos;
+        uint64_t pos;
+    };
+    static std::multimap<uint256, unknownParentPos> mapBlocksUnknownParent;
```

Comment:
Does it bother anyone else that this is `static`? This is global state, which, in general, equals evil. It must be static, given that it's declared here, since its state must persist across calls to this function (very often, a block and its parent are in different `blk` files). We could eliminate this global state by allocating this map in the caller of this function, `src/init.cpp: ThreadImport()`, and passing it by reference to `LoadExternalBlockFile()`. 

Another thing that seems missing from the current design is, what happens if there are leftover entries in this map after reindexing has completed? That means we've read blocks from disk but never found their parent blocks. In my testing, this map is empty at the end of reindexing, as expected. But what if it wasn't? That would mean we're missing one or more `blocks/blk00nnn.dat` files, or some blocks from those files. Isn't that worth at least a `LogPrintf()`? Making this change would also require moving the map out to the caller, because this function doesn't know if it will be called again.

Reply:
UPDATE: this was fixed in #25571